src/_P021_Level.ino

```diff
@@ -182,93 +389,465 @@ boolean Plugin_021(uint8_t function, struct EventStruct *event, String& string)
+  // - The logical state of the output signal [on,OFF]
+  // - The state of the internal state machine
+  // - Timestamp for last change of the output signal
+  // Note: the actual state of the output is not stored, it can be calculated from the control_state
+  UserVar.setFloat(event->TaskIndex, P021_VALUE_OUTPUT, (float)(((P021_control_state)new_control_state == P021_STATE_IDLE) ? 0 : 1));
+  UserVar.setFloat(event->TaskIndex, P021_VALUE_STATE,  (float)new_control_state);
+  P021_timestamp[event->TaskIndex] = timestamp;
+
+  if (new_control_state != old_control_state)
+  {
+    sendData(event);
+  }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Input evaluation to determine switching the control signal on
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+bool P021_check_on(float value, float setpoint, float hysteresis, bool invert, bool symetric, bool force)
+{
+  if (force)
+  {
+    return true;
+  }
+
+  if (symetric)
+  {
+    bool  isZero             = essentiallyZero(hysteresis);
+    float valueLowThreshold  = setpoint - (isZero ? 0.0f : (hysteresis / 2.0f));
+    float valueHighThreshold = setpoint + (isZero ? 1.0f : (hysteresis / 2.0f)); // Include setvalue on 0-hysteresis
+    return invert ? !definitelyGreaterThan(value, valueLowThreshold) : !definitelyLessThan(value, valueHighThreshold);
+  }
+  else
+  {
+    return invert ? (value < setpoint) : (value > setpoint);
+  }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Input evaluation to determine switching the control signal off
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+bool P021_check_off(float value, float setpoint, float hysteresis, bool invert, bool symetric, bool force)
+{
+  if (force)
+  {
+    return false;
+  }
+
+  if (symetric)
+  {
+    bool  isZero             = essentiallyZero(hysteresis);
+    float valueLowThreshold  = setpoint - (isZero ? 0.0f : (hysteresis / 2.0f));
+    float valueHighThreshold = setpoint + (isZero ? 1.0f : (hysteresis / 2.0f)); // Include setvalue on 0-hysteresis
+    return invert ? !definitelyLessThan(value, valueHighThreshold) : !definitelyGreaterThan(value, valueLowThreshold);
+  }
+  else
+  {
+    return invert ? (value > (setpoint + hysteresis)) : (value < (setpoint - hysteresis));
+  }
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Retrieve one bit flag from a set of flags stored in one UINT16
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+boolean P021_getConfigBit(uint16_t config, int pos)
+{
+  boolean res = ((config >> pos) & 0x0001) == 1;
+
+  return res;
+}
```

Comment:
Arduino provides a `bitRead()` function/macro, no need to re-invent that wheel.

Reply:
Thanks, replaced.